node/src/main/scala/coop/rchain/node/node.scala

```diff
@@ -314,13 +301,74 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
-      case th if th.containsMessageWith("Error loading shared library libsodium.so") =>
-        Log[Task]
-          .error(
-            "Libsodium is NOT installed on your system. Please install libsodium (https://github.com/jedisct1/libsodium) and try again.")
-      case th =>
-        th.getStackTrace.toList.traverse(ste => Log[Task].error(ste.toString))
-    } *> exit0.as(Right(())))
+  /**
+    * Handles unrecoverable errors in program. Those are errors that should not happen in properly
+    * configured enviornment and they mean immediate termination of the program
+    */
+  private def handleUnrecoverableErrors(prog: Effect[Unit])(implicit log: Log[Task]): Effect[Unit] =
+    EitherT[Task, CommError, Unit](
+      prog.value
+        .onErrorHandleWith {
+          case th if th.containsMessageWith("Error loading shared library libsodium.so") =>
+            Log[Task]
+              .error(
+                "Libsodium is NOT installed on your system. Please install libsodium (https://github.com/jedisct1/libsodium) and try again.")
+          case th =>
+            th.getStackTrace.toList.traverse(ste => Log[Task].error(ste.toString))
+        } *> exit0.as(Right(())))
+
+  private def generateCasperConstructor(runtimeManager: RuntimeManager)(
+      implicit
+      log: Log[Task],
+      time: Time[Task],
+      transport: TransportLayer[Task],
+      nodeDiscovery: NodeDiscovery[Task],
+      blockStore: BlockStore[Effect],
+      oracle: SafetyOracle[Effect]): Effect[MultiParentCasperConstructor[Effect]] =
+    MultiParentCasperConstructor.fromConfig[Effect, Effect](conf.casperConf, runtimeManager)
+
+  /**
+    * Main node entry. Will Create instances of typeclasses and run the node program.
+    */
+  val main: Effect[Unit] = for {
+
+    /** create typeclass instances */
+    connectionsState <- effects.connectionsState[Task].pure[Effect]
+    log              = effects.log
+    time             = effects.time
+    sync             = SyncInstances.syncEffect
+    transport = effects.tcpTransportLayer(host, port, certificateFile, keyFile)(src)(
+      scheduler,
+      connectionsState,
+      log)
+    kademliaRPC = effects.kademliaRPC(src, defaultTimeout)(metricsTask, transport)
+    nodeDiscovery = effects.nodeDiscovery(src, defaultTimeout)(log,
+                                                               time,
+                                                               metricsTask,
+                                                               transport,
+                                                               kademliaRPC)
+    blockStore     = LMDBBlockStore.create[Effect](conf.casperBlockStoreConf)(sync, metrics)
+    _              <- blockStore.clear() // FIX-ME replace with a proper casper init when it's available
+    oracle         = SafetyOracle.turanOracle[Effect](Applicative[Effect], blockStore)
+    runtime        <- Runtime.create(storagePath, storageSize).pure[Effect]
```

Comment:
There is an inconsistency here between `<- ...pure[Effect]` and `=`. Please fix in a future PR.

Reply:
👍 roger that!